### PR TITLE
(DOCS-4304) Go tracing guides

### PR DESCRIPTION
**Pull request by hestonhoffman**
### What does this PR do?
Adds Go APM Tutorials

### Motivation
DOCS-4304

---

### Reviewer checklist
- [ ] Review the changed files.
- [ ] Review the URLs listed in the [Preview](#preview) section.
- [ ] Check images for PII
- [ ] Review any mentions of "Contact Datadog support" for internal support documentation.

### CODE REVIEW

**Comment by jonbodner** [comment]
Under "Troubleshooting", it says "Java tracer". It should say "Go tracer" and the link on "Enable debug mode" should be to https://docs-staging.datadoghq.com/heston/go-tracing-guides/tracing/troubleshooting/tracer_debug_logs/?code-lang=go .

**Comment by jonbodner** [comment]
The text:

```
To install the necessary libraries for the sample app, run the following command from the root directory of the repo:

go mod download
``` 

Should be deleted. This is not needed.

**Comment by jonbodner** [comment]
change the first `make run` to `make runNotes`

**Comment by jonbodner** [comment]
change the first `make exit` to `make exitNotes`

**Comment by jonbodner** [comment]
Change "Ensure that the following lines of code are present in each of the apps defined in apm-tutorial-golang/cmd/calendar/main.go and apm-tutorial-golang/cmd/notes/main.go:" to 

```
Uncomment the following lines of code in apm-tutorial-golang/cmd/calendar/main.go
```

**Comment by jonbodner** [comment]
Delete the code block:

func main() {
  tracer.Start()
  defer tracer.Stop()
  ...
}

and add the following lines before "Launch the Go application and explore automatic instrumentation":

Now that the tracing library has been added to go.mod, we are now going to enable tracing support in apm-tutorial-golang/cmd/calendar/main.go. 

Uncomment the following imports:

```
	sqltrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql"
	chitrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/go-chi/chi"
	httptrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/net/http"
	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
```

In main(), uncomment the lines:
```
	tracer.Start()
	defer tracer.Stop()
```

and

```
	client = httptrace.WrapClient(client, httptrace.RTWithResourceNamer(func(req *http.Request) string {
		return fmt.Sprintf("%s %s", req.Method, req.URL.Path)
	}))
```

and

```
	r.Use(chitrace.Middleware(chitrace.WithServiceName("notes")))
```

In setupDB(), uncomment the lines:
```
	sqltrace.Register("sqlite3", &sqlite3.SQLiteDriver{}, sqltrace.WithServiceName("db"))
	db, err := sqltrace.Open("sqlite3", "file::memory:?cache=shared")
```

and comment out the line:

```
	db, err := sql.Open("sqlite3", "file::memory:?cache=shared")
```

**Comment by jonbodner** [comment]
Change the code block:

```
go get gopkg.in/DataDog/dd-trace-go.v1/ddtrace
go get gopkg.in/DataDog/dd-trace-go.v1/internal/remoteconfig@v1.46.1
```

to:

```
go get gopkg.in/DataDog/dd-trace-go.v1/ddtrace
go mod tidy
```

**Comment by jonbodner** [comment]
Add before  "Launch the Go application and explore automatic instrumentation" (but after all the uncomment stuff):

Once you have made these changes, run the command:

```
go mod tidy
```

**Comment by jonbodner** [comment]
Change the paragraph:

The makeSpanMiddleware function starts a span using the context found from the HTTP request. It sets the name of the span to the function name used to resolve the REST call:



To:

Uncomment the `makeSpanMiddleware` function in `notes/notesController.go`. It generates middleware that wraps a request in a span with the supplied name. To use this function, comment out the lines:

```
	r.Get("/notes", nr.GetAllNotes)                // GET /notes
	r.Post("/notes", nr.CreateNote)                // POST /notes
	r.Get("/notes/{noteID}", nr.GetNoteByID)       // GET /notes/123
	r.Put("/notes/{noteID}", nr.UpdateNoteByID)    // PUT /notes/123
	r.Delete("/notes/{noteID}", nr.DeleteNoteByID) // DELETE /notes/123
```

and uncomment the lines:

```
	r.Get("/notes", makeSpanMiddleware("GetAllNotes", nr.GetAllNotes))               // GET /notes
	r.Post("/notes", makeSpanMiddleware("CreateNote", nr.CreateNote))                // POST /notes
	r.Get("/notes/{noteID}", makeSpanMiddleware("GetNote", nr.GetNoteByID))          // GET /notes/123
	r.Put("/notes/{noteID}", makeSpanMiddleware("UpdateNote", nr.UpdateNoteByID))    // PUT /notes/123
	r.Delete("/notes/{noteID}", makeSpanMiddleware("DeleteNote", nr.DeleteNoteByID)) // DELETE /notes/123
```

also uncomment the import:

```
	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
```

**Comment by jonbodner** [comment]
Before

For more information on custom tracing

add the line:

Uncomment these lines to create these spans.

**Comment by jonbodner** [comment]
After "When it is set to y, the notes application calls the calendar application to get a date to add to the note.", add the lines:

To enable tracing in the calendar application, uncomment the following lines in `cmd/calendar/main.go`:

```
	chitrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/go-chi/chi"
	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
```

```
	tracer.Start()
	defer tracer.Stop()
```

and

```
	r.Use(chitrace.Middleware(chitrace.WithServiceName("calendar")))
```

1. If the notes application is still running, use `make exitNotes` to stop it.

Then change the existing 1. to:

2. run `make run` to start the sample applications.

**Comment by jonbodner** [comment]
In https://docs-staging.datadoghq.com/heston/go-tracing-guides/tracing/guide/tutorial-enable-go-host/#install-datadog-tracing , the line:

Uncomment the following imports in apm-tutorial-golang/cmd/calendar/main.go:

Should be:

Uncomment the following imports in apm-tutorial-golang/cmd/notes/main.go:

**Comment by jonbodner** [comment]
In the list of imports under "Uncomment the following imports in apm-tutorial-golang/cmd/calendar/main.go:", add the import "fmt":

```
	sqltrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql"
	chitrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/go-chi/chi"
	httptrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/net/http"
	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
        "fmt"
````

**Comment by jonbodner** [comment]
Under Use custom tracing with context on the go host guide, the first line in the source code is indented incorrectly. It should line up with everything else.

**Comment by jonbodner** [comment]
Before the line "For more information on custom tracing, see [Go Custom Instrumentation](https://docs-staging.datadoghq.com/heston/go-tracing-guides/tracing/trace_collection/custom_instrumentation/go/).", insert the following:

Uncomment the following imports:

```
	// "gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
	// "gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
```

**Comment by jonbodner** [comment]
The source code block headers in the section [Examine distributed traces](https://docs-staging.datadoghq.com/heston/go-tracing-guides/tracing/guide/tutorial-enable-go-host/#examine-distributed-traces) are wrong. They should say "cmd/calendar/main.go". 

Also, the first line of each code block is indented incorrectly. They should line up with the other source code in the block.

**Review by kayayarai — CHANGES_REQUESTED** [review]
Just a few little things. I think most of the comments in `host` also apply to the others, but I just commented once.

File: content/en/tracing/guide/tutorial-enable-go-host.md [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+
+Uncomment the following imports:
+
+{{< code-block lang="go" filename="notes/notesHelper.go" disable_copy="true" collapsible="true" >}}
+  "gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+  "gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+{{< /code-block >}}
+
+For more information on custom tracing, see [Go Custom Instrumentation][12].
+
```
[the remark this answers is not in the document]
Comment (hestonhoffman) [reply content/en/tracing/guide/tutorial-enable-go-host.md:309]: I should probably include an image as well for reference. I'll work on adding this in.

**Review by kayayarai — APPROVED** [review]

**Pull request merged**